### PR TITLE
feat(ml): torch db stores encoded images instead of tensors

### PR DESCRIPTION
**Pull request by beniz**
This PR defaults torch db for image input connector to store encoded images instead of torch tensors. This allows a 20x compression ratio of the db size, and does not appear to affect prefetching speed while training.

### CODE REVIEW

File: src/backends/torch/torchdataset.h [hunk 1]
```diff
@@ -79,6 +79,10 @@ namespace dd
         = nullptr;               /**< back ptr to input connector. */
     bool _classification = true; /**< whether a classification dataset. */
 
+    std::default_random_engine _rnd_generator;
```
Comment (BynaryCobweb) [inline src/backends/torch/torchdataset.h:82]: where is this used?

**Review by fantes — APPROVED** [review]

File: src/backends/torch/torchdataset.cc [hunk 2]
```diff
@@ -116,6 +116,96 @@ namespace dd
       }
   }
 
+  void TorchDataset::write_image_to_db(const cv::Mat &bgr,
+                                       const torch::Tensor &target)
+  {
+    // serialize image
+    std::stringstream dstream;
+    std::vector<uint8_t> buffer;
+    cv::imencode(".jpg", bgr, buffer);
```
[the remark this answers is not in the document]
Comment (beniz) [reply src/backends/torch/torchdataset.cc:125]: The images would be slightly different yes. One way to alleviate this is to set quality to 100% (default is 95% in OpenCV).

```
 std::vector<int> param(2);
param[0] = cv::IMWRITE_JPEG_QUALITY;
param[1] = 100;
cv::imencode(".jpg", bgr, buffer, param);
```

File: src/backends/torch/torchdataset.h (hunk 1 above)
Comment (beniz) [reply src/backends/torch/torchdataset.h:82]: Good catch, it's a leak from my data augmentation branch, will remove.

File: src/backends/torch/torchdataset.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (beniz) [reply src/backends/torch/torchdataset.cc]: Added.

File: src/backends/torch/torchdataset.h (hunk 1 above)
Comment (beniz) [reply src/backends/torch/torchdataset.h]: Done.

File: src/backends/torch/torchdataset.cc [hunk 3]
```diff
@@ -116,6 +116,96 @@ namespace dd
...
+      }
+  }
+
+  void TorchDataset::read_image_from_db(const std::string &datas,
+                                        const std::string &targets,
+                                        cv::Mat &bgr, torch::Tensor &targett)
+  {
+    std::vector<uint8_t> img_data(datas.begin(), datas.end());
+    bgr = cv::Mat(img_data, true);
+    bgr = cv::imdecode(bgr, cv::IMREAD_COLOR); // TODO: bw
```
[the remark this answers is not in the document]
Comment (beniz) [reply src/backends/torch/torchdataset.cc]: Done.

**Comment by BynaryCobweb** [comment]
It would be even better if we could put the image in db without even decoding it (directly from disk), but this can be done in a future PR.

**Review by BynaryCobweb — APPROVED** [review]

**Pull request merged**